File: backend/src/main/java/com/ems/employee_management_system/utils/HashUtil.java
package com.ems.employee_management_system.utils;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

/**
 * Utility class for hashing operations
 * Matches frontend hashing implementation
 */
public final class HashUtil {
    
    // Private constructor to prevent instantiation
    private HashUtil() {
        throw new UnsupportedOperationException("Utility class cannot be instantiated");
    }
    
    private static final String SALT_PREFIX_PASSWORD = "ems_";
    private static final String SALT_SUFFIX_PASSWORD = "_salt";
    private static final String SALT_PREFIX_USERNAME = "ems_user_";
    
    /**
     * Hash a string using SHA-256
     * @param input The string to hash
     * @return Hex-encoded SHA-256 hash
     */
    public static String hashSHA256(String input) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(input.getBytes(StandardCharsets.UTF_8));
            
            // Convert bytes to hex string
            StringBuilder hexString = new StringBuilder();
            for (byte b : hash) {
                String hex = Integer.toHexString(0xff & b);
                if (hex.length() == 1) {
                    hexString.append('0');
                }
                hexString.append(hex);
            }
            return hexString.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException("SHA-256 algorithm not available", e);
        }
    }
    
    /**
     * Hash password using the same method as frontend
     * @param password Plain text password
     * @return Hashed password
     */
    public static String hashPassword(String password) {
        String saltedPassword = SALT_PREFIX_PASSWORD + password + SALT_SUFFIX_PASSWORD;
        return hashSHA256(saltedPassword);
    }
    
    /**
     * Hash username/user ID using the same method as frontend
     * @param userId Username or user identifier
     * @return Hashed user ID
     */
    public static String hashUserId(String userId) {
        String saltedUserId = SALT_PREFIX_USERNAME + userId;
        return hashSHA256(saltedUserId);
    }
    
    /**
     * Check if a string looks like a hash (hex string of 64 characters for SHA-256)
     * @param input String to check
     * @return true if it looks like a hash
     */
    public static boolean isHash(String input) {
        if (input == null || input.length() != 64) {
            return false;
        }
        // Check if it's a valid hex string
        return input.matches("[0-9a-f]{64}");
    }
}

File: frontend/src/app/core/utils/hash.util.ts
/**
 * Utility for hashing sensitive data before sending to backend
 * Uses Web Crypto API for secure hashing (with crypto-js fallback for HTTP contexts)
 */

import * as CryptoJS from 'crypto-js';

/**
 * Hash a string using SHA-256
 * Uses Web Crypto API if available (HTTPS/localhost), otherwise uses crypto-js fallback
 * @param data The string to hash
 * @returns Promise resolving to the hex-encoded hash
 */
export async function hashString(data: string): Promise<string> {
  // Check if Web Crypto API is available (requires HTTPS or localhost)
  if (typeof crypto !== 'undefined' && crypto.subtle) {
    try {
      // Convert string to ArrayBuffer
      const encoder = new TextEncoder();
      const dataBuffer = encoder.encode(data);

      // Hash using Web Crypto API
      const hashBuffer = await crypto.subtle.digest('SHA-256', dataBuffer);

      // Convert ArrayBuffer to hex string
      const hashArray = Array.from(new Uint8Array(hashBuffer));
      const hashHex = hashArray.map(b => b.toString(16).padStart(2, '0')).join('');

      return hashHex;
    } catch (error) {
      console.warn('Web Crypto API failed, using crypto-js fallback:', error);
      // Fall through to fallback implementation
    }
  }

  // Fallback: Use crypto-js for HTTP contexts (works everywhere)
  return CryptoJS.SHA256(data).toString(CryptoJS.enc.Hex);
}

/**
 * Hash password before sending to backend
 * @param password Plain text password
 * @returns Promise resolving to hashed password
 */
export async function hashPassword(password: string): Promise<string> {
  // Add a salt/prefix to make it harder to reverse (though backend will hash again)
  const saltedPassword = `ems_${password}_salt`;
  return hashString(saltedPassword);
}

/**
 * Hash user identifier (username) before sending to backend
 * @param userId Username or user identifier
 * @returns Promise resolving to hashed user ID
 */
export async function hashUserId(userId: string): Promise<string> {
  const saltedUserId = `ems_user_${userId}`;
  return hashString(saltedUserId);
}

/**
 * Hash multiple values and combine them
 * @param values Array of strings to hash
 * @returns Promise resolving to combined hash
 */
export async function hashMultiple(...values: string[]): Promise<string> {
  const combined = values.join('|');
  return hashString(combined);
}

